lightning/src/routing/scoring.rs

```diff
@@ -1233,76 +1233,84 @@ impl<G: Deref<Target = NetworkGraph<L>>, L: Deref, T: Time> Score for Probabilis
 			.saturating_add(base_penalty_msat)
 	}
 
-	fn payment_path_failed(&mut self, path: &[&RouteHop], short_channel_id: u64) {
-		let amount_msat = path.split_last().map(|(hop, _)| hop.fee_msat).unwrap_or(0);
+	fn payment_path_failed(&mut self, path: &Path, short_channel_id: u64) {
+		let amount_msat = path.final_value_msat();
 		log_trace!(self.logger, "Scoring path through to SCID {} as having failed at {} msat", short_channel_id, amount_msat);
 		let network_graph = self.network_graph.read_only();
-		for (hop_idx, hop) in path.iter().enumerate() {
-			let target = NodeId::from_pubkey(&hop.pubkey);
+		for (hop_idx, (pubkey, scid)) in path.hops.iter().map(|hop| (hop.pubkey, hop.short_channel_id))
+			.chain(path.blinded_tail.as_ref().map(|tail| (tail.path.introduction_node_id, tail.intro_node_scid)))
```

Comment:
Maybe we can DRY the iterator formation in a helper?

Reply:
Took a stab at this. I followed the compiler's suggestion to make the return value of the helper `Chain<Map<Iter<'_, RouteHop>, [closure@lightning/src/routing/router.rs:323:24: 323:64]>, IntoIter<(PublicKey, u64)>>`, except I used `Fn` for the closure. 

That got me:
```
error[E0277]: expected a `Fn<(&RouteHop,)>` closure, found `[closure@lightning/src/routing/router.rs:323:24: 323:64]`
```

So then I tried to make the helper take a callback, and ran into:
```
error[E0500]: closure requires unique access to `self` but it is already borrowed
    --> lightning/src/routing/scoring.rs:1240:26
     |
1239 |         let network_graph = self.network_graph.read_only();
     |                             ------------------------------ borrow occurs here
1240 |         path.node_id_scid_iter(|(hop_idx, (pubkey, scid))| {
     |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^ closure construction occurs here
...
1255 |                     self.channel_liquidities
     |                     ------------------------ second borrow occurs due to use of `self` in closure
...
1306 |     }
     |     - first borrow might be used here, when `network_graph` is dropped and runs the destructor for type `ReadOnlyNetworkGraph<'_>`
```
We could maybe `Mutex` `channel_liquidities` within the scorer, which should allow us to go down the callback route by avoiding a mutable second borrow of `self`, I’d rather not do that in this PR, though.

Reply:
Could you have the helper return an `impl Iterator` instead?

Reply:
Thanks, not sure how I missed that